Add apatite fission track module

ptrscll opened the pull request:
This PR contains the aft module and the corresponding tests. As of right now, it is still a work in progress, but I figured I'd make this PR in case you would like to look at my progress before our meeting tomorrow. 

So far, I have made 2 changes to constants from the old code:
1. For converting between years and seconds, I updated the number of days in a year from 365.25 to 365.2422. This had a negligible effect on the results of all Ketcham (2005) tests. 
2. I realized the constant 0.9157 in r_to_rho had a typo and was written as 0.9175. I corrected this typo, which only affected age calculations and not length calculations. The results of the change are given below:

Ketcham (2005) Test | Age from Old Code (Ma) | Age from New Code (Ma) | HeFTy Age (Ma)
--|--|--|--
Test 1                        | 89.3                                          | 89.3                                                  | 89.4
Test 2 | 39.9 | 40.0 | 39.8
Test 3 | 62.6 | 63.3 | 63.3
Test 4 (Dpar=1.75) | 13.1 | 13.2 | 13.0
Test 4 (Dpar=2.50) | 50.8 | 51.6 | 51.1

ptrscll commented:
Okay, this code should be ready for review! I also want to highlight two aspects of the code that I'm a bit uncertain about:

1. Does the way I've setup how constants are passed to functions make sense? I've tried to make constants that come from Ketcham passed as dictionaries and make other constants parameters to their respective functions. I'm not sure if the dictionary approach makes the most sense, since on the one hand, it makes it annoying to manually modify specific constants, but on the other hand, it reduces the number of extra parameters passed to functions and helps indicate where some of these constants are coming from.
2. I'm not sure exactly how we want to test the length distributions output by the forward model, given that they still don't align nicely with HeFTy output. Right now, I just have one assert statement in the forward model test related to the mean of one of the simpler length distributions, which I don't think is enough. Two potential approaches could either be using dummy input data nad constants (which would end up getting rather complicated to write because of all the intermediate calculations involved in obtaining the final distribution) or checking that the most frequent lengths line up between our model and HeFTy (since that particular value tends to line up well). Let me know what your thoughts are here. 

Also, when I tried squashing commits (on GitHub Desktop), GitHub wouldn't let me squash most of the commits because they pre-dated a merge commit. I'm not sure if there's a way around this, but I've squashed at least the commits I was able to squash.

dyvasey reviewed (commented):
I think this looks really good overall! In the review comments below, the only systematic stylistic thing is to slightly formalize the returns in the docstrings so they match the parameters - even if there isn't a specific variable, do the first line as the type and then the second as the indented description.

Responses to specific issues:

1. I think the constants are fine for now - I think anyone who is going to want to modify individual model parameters can also deal with creating a dictionary in their script.
2. I spent some time looking over the length distributions code and some of the Ketcham papers - I think doing tests that the main peaks (i.e., the modes) approximately line up is good for now. I think this must be related to the relationship between mean track length and standard deviation - we're using a static empirical equation from Ketcham (2005), and HeFTy may have a more sophisticated empirical approach using multiple datasets. Would be great to improve this, but for now, the general shape is right (which is a model after all), and the main goal right now is the age prediction, which is unaffected.
3. For squashing commits, I would suggest doing interactive rebase from the command line locally:
`git rebase -i HEAD~X` where X is the number of commits you want to modify. That will load a text file where you can squash commits (by changing `pick` to `s`) and/or reorder them. It may create merge conflicts, but you can resolve those locally and then force push the whole thing.

See here: https://docs.github.com/en/get-started/using-git/using-git-rebase-on-the-command-line

dyvasey commented on gdtchron/aft.py:
```diff
@@ -0,0 +1,656 @@
+    """
+    starts = tsteps[:-1] / 1e6  # t2 (Myr)
+    ends = tsteps[1:] / 1e6     # t1 (Myr)
+    return (np.exp(lamb * starts) - np.exp(lamb * ends)) / lamb
+
```
or similar, I may not have this quite right.

ptrscll replied:
I just made this modification to a local version of thesis-scully, and this seems to work a lot better! Some of the standard deviations still don't line up super nicely, but the means are now much closer. Here's a table of results

Ketcham (2005) Test | Old Mean (micrometers) | New Mean | HeFTy Mean | Old Stdev | New Stdev | HeFTy Stdev
--|--|--|--|--|--|--
Test 1|15.02|15.03|15.03|0.61|0.58|0.82
Test 2 |13.70|14.24|14.24|1.81|1.24|1.35
Test 3 |13.05|13.39|13.38|1.44|1.47|1.56
Test 4 (Dpar=1.75) | 12.31 | 13.93 | 13.95|2.62|1.70|1.70
Test 4 (Dpar=2.50) | 12.33 | 12.88 | 12.91|1.60|1.50|1.52

I can work on updating the gdtchron code to include this weighing scheme this weekend.


Also, here's the new length distributions:

![thesis_AFT_figure](https://github.com/user-attachments/assets/959fe4fe-c92d-479d-9272-245d2b173696)

(vs. the old ones, for comparison):
![fig5](https://github.com/user-attachments/assets/90539f87-e4a2-45a0-9540-573ebf5f5481)

dyvasey replied:
Awesome! Agreed that this looks a lot better and would be good to implement.

ptrscll commented on gdtchron/aft.py:
```diff
@@ -0,0 +1,656 @@
+        mean of the mixed distribution; larger values produce smoother curves
+        when plotting the frequency series. Only relevant if get_lengths is True
+
+    Returns
```
Should I keep some of the logical structure because we only return a tuple right now if the user sets `get_lengths` to true? Or should we just return a tuple containing the age if `get_lengths` is false?

dyvasey replied:
Fair point - maybe do list the individual returns but just list all except age as optional? I think it'll look cleaner if we can keep the if statement out of the returns and avoid listing age twice.

dyvasey replied:
Generally good! I would remove the overarching `final_results` from `forward_model_aft`, since that isn't a named variable and, as you point out, in some conditions the return isn't a tuple. So I'm going back on my original comment there and saying just name the individual parts of the return, but mark all except age optional.

`calc_l_dist` is good since `results` is a named variable and is always a tuple.

ptrscll commented:
Okay, I just updated `calc_weights`. I switched from using the equation you wrote up for `w2` to using `r_to_rho` (which takes from the midpoints of each timestep and removes r that are too small to be reasonably observed), but that only changed a couple means/standard deviations by 0.01-0.02. I still need to write the length distribution tests for forward_model_aft and squash the commits, but other than that I've made all the changes we've talked about.

The one other thing I noticed while I was looking back at the HeFTy data is that the way we determine length values for our length distribution is slightly different from HeFTy. Our approach currently just gives us a range of values based on what's two standard deviations away from the means of the smallest and largest distributions, but, looking at the raw HeFTy data, HeFTy always just uses a fixed distribution from 0 to 20 micrometers with an interval of 0.25 micrometers between each length. I can definitely switch to following HeFTy here, but I wanted to double check with you first, since our current approach is still technically correct (and produces nearly identical length means/standard deviations). If we do switch over, I think we should keep the parameter letting users change the number of lengths used in for the distribution, since using only 81 (the number HeFTy uses) can produce some less smooth graphs (see below).
[testing.pdf](https://github.com/user-attachments/files/20640236/testing.pdf)

dyvasey commented:
Looks good! I think this is good to go once the tests are done commits have been reorganized.

> Our approach currently just gives us a range of values based on what's two standard deviations away from the means of the smallest and largest distributions, but, looking at the raw HeFTy data, HeFTy always just uses a fixed distribution from 0 to 20 micrometers with an interval of 0.25 micrometers between each length. I can definitely switch to following HeFTy here, but I wanted to double check with you first, since our current approach is still technically correct (and produces nearly identical length means/standard deviations). If we do switch over, I think we should keep the parameter letting users change the number of lengths used in for the distribution, since using only 81 (the number HeFTy uses) can produce some less smooth graphs (see below).

I don't feel strongly about this if it's not changing the actual distributions significantly. It's really just a question of visualization. So I'm fine with sticking with what we have.

dyvasey commented:
Yep! We should add `from gdmate import aft` and add "aft" to the list `__ALL__`. This way if someone runs `import gdmate` they will also import `gdmate.aft`

ptrscll commented:
Okay, `__init__.py` should be updated and everything should be squashed now!
